#01 importing libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

#02 creating a database

In [2]:
df_test=pd.DataFrame()

In [3]:
df_test['mix']=['a','b',1,True]

In [4]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [5]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [6]:
#changing all datatypes to string
df_test['mix'] = df_test['mix'].astype('str')

#03 importing databases

In [7]:
path = r'C:\Users\arsla\Instacart basket analysis July 2025\02 Data'

In [8]:
df_prods = pd.read_csv(os.path.join(path,'Original Data','products.csv'))

In [9]:
#finding null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
#creating a dataframe that contains only null values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# removing columns that have missing values 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
df_prods.shape

(49693, 5)

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
#finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

#04 beginning of the exercise 4.4

TASKQ2

In [20]:
df_ords = pd.read_csv(os.path.join(path,'Original Data','orders.csv'))

In [21]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [22]:
df_ords_prep = pd.read_csv(os.path.join(path,'Prepared Data','orders_wrangled.csv'))

In [23]:
df_ords_prep.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

days_since_prior_order column has fewer count than the others which indicates missing values.

TASKQ3

In [24]:
for col in df_ords_prep.columns.tolist():
  weird = (df_ords_prep[[col]].map(type) != df_ords_prep[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_prep[weird]) > 0:
    print (col)

In [25]:
df_ords_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


#there isn't mixe-type data

TASKQ5

In [26]:
#finding missing values
df_ords_prep.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in days_since_prior_order column

TASKQ6

In [27]:
df_ords_prep_nan=df_ords_prep[df_ords_prep['days_since_prior_order'].isnull()==True]

In [28]:
df_ords_prep_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Used this method beacuse it is safer to create a new dataframe rather than overwrite the original one. Then I'll impute values into missing values I'll impute median value since it feels the correct way to do it

In [29]:
#imputing values
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].fillna(7)

In [30]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     7.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [31]:
df_ords_prep

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [32]:
df_ords_prep_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

TASKQ7

In [33]:
#checking for duplicates
df_ords_dups=df_ords_prep_nan[df_ords_prep_nan.duplicated()]

In [34]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There isn't any duplicate 

TASKQ9

In [35]:
df_ords_prep_nan.to_csv(os.path.join(path,'Prepared Data', 'orders_checked.csv'))

In [36]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data','products_checked.csv'))

In [37]:
df_ords_prep_nan.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
11           11   2168274        2             1                   2   
26           26   1374495        3             1                   1   
39           39   3343014        4             1                   6   
45           45   2717275        5             1                   3   
50           50   2086598        6             1                   5   
54           54   2565571        7             1                   3   
75           75    600894        8             1                   6   
79           79    280530        9             1                   1   
83           83   1224907       10             1                   2   
89           89   2411567       11             1                   0   
97           97   1483446       12             1                   1   
103         103   2618231       13             1                   6   
116         116   2906263       14             1                   2   
130         130   2514324       15             1                   5   
153         153   1273676       16             1                   2   
160         160   1737705       17             1                   2   
201         201   2780889       18             1                   6   
208         208   1321974       19             1                   6   
218         218    947722       20             1                   1   
223         223   2625565       21             1                   1   
257         257   2300537       22             1                   3   
273         273   1792521       23             1                   5   
278         278    611214       24             1                   3   
297         297   2610576       25             1                   1   
301         301   3229714       26             1                   2   
314         314    361493       27             1                   3   
396         396   1138968       28             1                   0   
421         421   2064090       29             1                   5   
440         440    375303       30             1                   1   

     order_hour_of_day  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
50                  18                     NaN  
54                   9                     NaN  
75                   0                     NaN  
79                  17                     NaN  
83                  14                     NaN  
89                  11                     NaN  
97                  12                     NaN  
103                 12                     NaN  
116                 18                     NaN  
130                 13                     NaN  
153                  8                     NaN  
160                 13                     NaN  
201                 18                     NaN  
208                 13                     NaN  
218                 10                     NaN  
223                 10                     NaN  
257                  7                     NaN  
273                 14                     NaN  
278                 20                     NaN  
297                 14                     NaN  
301                 15                     NaN  
314                  9                     NaN  
396                 14                     NaN  
421                 11                     NaN  
440                 13                     NaN

In [38]:
df_ords_prep.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   
11          11   2168274        2             1                   2   
12          12   1501582        2             2                   5   
13          13   1901567        2             3                   1   
14          14    738281        2             4                   2   
15          15   1673511        2             5                   3   
16          16   1199898        2             6                   2   
17          17   3194192        2             7                   2   
18          18    788338        2             8                   1   
19          19   1718559        2             9                   2   
20          20   1447487        2            10                   1   
21          21   1402090        2            11                   1   
22          22   3186735        2            12                   1   
23          23   3268552        2            13                   4   
24          24    839880        2            14                   3   
25          25   1492625        2            15                   1   
26          26   1374495        3             1                   1   
27          27    444309        3             2                   3   
28          28   3002854        3             3                   3   
29          29   2037211        3             4                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10                    30.0  
22                  9                    28.0  
23                 11                    30.0  
24                 10                    13.0  
25                 11                    30.0  
26                 14                     NaN  
27                 19                     9.0  
28                 16                    21.0  
29                 18                    20.0